catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/CatalogServiceImpl.java

```diff
@@ -303,42 +313,133 @@ public CompletionStage<SnapshotResponse> retrieveSnapshot(
+        icebergStuff().retrieveIcebergSnapshot(snapshotObjIdForContent(content), content);
 
-    ObjId snapshotId = snapshotObjIdForContent(content);
+    BiFunction<NessieEntitySnapshot<?>, List<NessieEntitySnapshot<?>>, SnapshotResponse>
+        responseBuilder =
+            (snapshot, history) ->
+                snapshotResponse(key, content, reqParams, snapshot, history, effectiveReference);
+
+    if (!reqParams.snapshotFormat().includeOldSnapshots()) {
+      return currentSnapshotStage.thenApply(snapshot -> responseBuilder.apply(snapshot, List.of()));
+    }
+
+    return currentSnapshotStage.thenCompose(
+        snapshot -> collectSnapshotHistory(snapshot, historyLog, responseBuilder));
+  }
+
+  private <R> CompletionStage<R> collectSnapshotHistory(
+      NessieEntitySnapshot<?> snapshot,
+      Iterator<ContentHistoryEntry> historyLog,
+      BiFunction<NessieEntitySnapshot<?>, List<NessieEntitySnapshot<?>>, R> responseBuilder) {
+    if (!(snapshot instanceof NessieTableSnapshot)) {
+      // Not an Iceberg table, no previous snapshots
+      return completedStage(responseBuilder.apply(snapshot, List.of()));
+    }
+
+    var tableSnapshot = (NessieTableSnapshot) snapshot;
+    var previousSnapshotIds = tableSnapshot.previousIcebergSnapshotIds();
+    if (previousSnapshotIds.isEmpty()) {
+      return completedStage(responseBuilder.apply(snapshot, List.of()));
+    }
+
+    // Collect history
+
+    var remainingSnapshotIds = new HashSet<>(previousSnapshotIds);
+
+    var collectorStage = completedStage(new SnapshotHistoryCollector(snapshot));
+
+    while (!remainingSnapshotIds.isEmpty() && historyLog.hasNext()) {
+      var next = historyLog.next();
+      var nextContent = next.getContent();
+      if (!(nextContent instanceof IcebergContent)) {
+        // should never happen, really
+        continue;
+      }
+      var nextSnapshotId = ((IcebergContent) nextContent).getVersionId();
+      if (!remainingSnapshotIds.remove(nextSnapshotId)) {
+        // not a snapshot that we need
+        continue;
+      }
 
-    CompletionStage<NessieEntitySnapshot<?>> snapshotStage =
-        icebergStuff().retrieveIcebergSnapshot(snapshotId, content);
+      var olderSnapStage =
+          icebergStuff()
+              .retrieveIcebergSnapshot(snapshotObjIdForContent(nextContent), nextContent)
+              .exceptionally(
+                  t -> {
+                    // There is not much we can do when the retrieval of table-metadata fails.
+                    LOGGER.warn(
+                        "Failed to retrieve table-metadata {}",
+                        ((IcebergContent) nextContent).getMetadataLocation());
+                    return null;
+                  });
+      collectorStage =
+          collectorStage.thenCombine(
+              olderSnapStage,
+              (collector, olderSnap) -> {
+                if (olderSnap != null) {
+                  collector.snapshotHistory.put(nextSnapshotId, olderSnap);
```

Comment:
I wonder whether we need the `collector` at all. The retrieval stages already know their position in the list and snapshot IDs. I wonder if it might be simpler to just put those stages into an array, do `CompletableFuture.allOf()` and then get all non-null results into the final list.

Reply:
Alternatively, given that `thenCombine` stages complete in order, we can probably just keep adding `olderSnap` directly to the list of results and pass it down to the next combined stage.

Reply:
SGTM, updated